client/client.go

```diff
@@ -129,38 +156,67 @@ func createClientWithClient(ctx context.Context, searchCli sserver1.SearchServer
 			return nil, err
 		}
 
-		indexer := libsearch.CreateSecureIndexBuilder(sha256.New, masterSecrets[i], tlfInfo.Salts, uint64(tlfInfo.Size))
-
-		dirInfo := DirectoryInfo{
-			absDir:      absDir,
-			tlfID:       tlfID,
-			indexer:     indexer,
-			pathnameKey: pathnameKey,
+		var indexers []*libsearch.SecureIndexBuilder
+		var pathnameKeys []libsearch.PathnameKeyType
+
+		// Sets up the indexers and pathname keys
+		if keyGen == libkbfs.PublicKeyGen {
+			masterSecret, err := fetchMasterSecret(directory, keyGen, lenMS)
+			if err != nil {
+				return nil, err
+			}
+			indexers = make([]*libsearch.SecureIndexBuilder, 1)
+			pathnameKeys = make([]libsearch.PathnameKeyType, 1)
+			indexers[0] = libsearch.CreateSecureIndexBuilder(sha256.New, masterSecret, tlfInfo.Salts, uint64(tlfInfo.Size))
+			copy(pathnameKeys[0][:], masterSecret[0:32])
+		} else if keyGen >= libkbfs.FirstValidKeyGen {
+			indexers = make([]*libsearch.SecureIndexBuilder, keyGen)
+			pathnameKeys = make([]libsearch.PathnameKeyType, keyGen)
+			for i := libkbfs.KeyGen(libkbfs.FirstValidKeyGen); i <= keyGen; i++ {
+				masterSecret, err := fetchMasterSecret(directory, i, lenMS)
+				if err != nil {
+					return nil, err
+				}
+				indexers[i-libkbfs.FirstValidKeyGen] = libsearch.CreateSecureIndexBuilder(sha256.New, masterSecret, tlfInfo.Salts, uint64(tlfInfo.Size))
```

Comment:
Might be cleaner to have a small helper function that just returns `i-libkbfs.FirstValidKeyGen` -- you can use it both here and below, and in `getKeyIndex()` above.